crates/aquascope/src/analysis/boundaries/mod.rs

```diff
@@ -96,6 +114,129 @@ fn select_candidate_location<'tcx>(
   }
 }
 
+/// Return the constraints that occur nested within a [`HirId`].
+///
+/// Note, constraints involving regions belonging to the same SCC are removed.
+fn flow_constraints_at_hir_id<'a, 'tcx: 'a>(
+  ctxt: &'a PermissionsCtxt<'a, 'tcx>,
+  ir_mapper: &'a IRMapper<'a, 'tcx>,
+  hir_id: HirId,
+) -> Option<Vec<(Origin, Origin, Point)>> {
+  let mir_locations =
+    ir_mapper.get_mir_locations(hir_id, GatherDepth::Nested)?;
+
+  let all_constraints = mir_locations
+    .values()
+    .flat_map(|loc| {
+      let ps = ctxt.location_to_points(loc);
+      ctxt
+        .polonius_input_facts
+        .subset_base
+        .iter()
+        .filter(move |&(f, t, p)| {
+          !ctxt.is_universal_subset((*f, *t)) && ps.contains(p)
+        })
+        .copied()
+    })
+    .collect::<Vec<_>>();
+
+  Some(all_constraints)
+}
+
+/// If flow permissions are enabled, find expected flow permissions (if any) for the
+/// given `hir_id` within the larger `flow_context`.
+fn get_flow_permission<'tcx>(
+  ctxt: &PermissionsCtxt<'_, 'tcx>,
+  ir_mapper: &IRMapper<'_, 'tcx>,
+  flow_context: HirId,
+  hir_id: HirId,
+  span_to_range: impl Fn(Span) -> Range,
+) -> Option<FlowBoundary> {
+  if !ENABLE_FLOW_PERMISSIONS
+    .copied()
+    .unwrap_or(ENABLE_FLOW_DEFAULT)
+  {
+    log::warn!("Flow permissions are disabled!");
+    return None;
+  }
+
+  let hir = ctxt.tcx.hir();
+  let body = &ctxt.body_with_facts.body;
+
+  let region_flows = ctxt.region_flows();
+
+  log::debug!(
+    "FLOW INFORMATION:\nBound {} ---\n{:#?}\nAt {} ---\n{:#?}",
+    hir.node_to_string(flow_context),
+    flow_constraints_at_hir_id(ctxt, ir_mapper, flow_context),
+    hir.node_to_string(hir_id),
+    flow_constraints_at_hir_id(ctxt, ir_mapper, hir_id)
+  );
+
+  // Do any given constraints have an abstract Origin on the RHS?
+  let has_abstract_on_rhs = |flows: &[(Origin, Origin, Point)]| {
+    flows
+      .iter()
+      .any(|&(_, t, _)| region_flows.has_abstract_member(t))
+  };
+
+  let context_constraints =
+    flow_constraints_at_hir_id(ctxt, ir_mapper, flow_context)?;
+
+  // FIXME: current restriction, only look at constraints when
+  // an abstract equivalent region is on the right-hand-side.
+  //
+  // This covers the cases:
+  // - missing abstract-outlives-abstract constraint.
+  // - local outlives abstract.
+  if !has_abstract_on_rhs(&context_constraints) {
+    return None;
+  }
+
+  log::debug!("flow context edges ---:\n{context_constraints:#?}");
+
+  // Search for relevant flows and flow violations.
+  let specific_constraints =
+    flow_constraints_at_hir_id(ctxt, ir_mapper, hir_id)?;
+  log::debug!("HirId local constraints ---:\n{specific_constraints:#?}");
+
+  let kind = context_constraints
+    .iter()
+    .find_map(|&(from, to, _)| {
```

Comment:
This finds the first violation, but would it ever be the case that there are multiple violations? If so, is arbitrarily picking one reasonable or would we want to heuristically decide which violation is important?

Reply:
I do see a sort of hierarchy in the violations. I would see the order of importance as:

1. local outlives abstract
2. missing outlives constraint
3. local invalidated at exit

I'm not sure which combination of those *could* occur together but reporting one seems sufficient for now because we aren't doing anything special visually. If we want to expand this in the future, it would be relatively easy but probably a decision driven by how we want the corresponding visualization to look.

Reply:
For now, let's just make this a `filter_map` and add another line explicitly selecting the first violation with a note about this.